### ISSUE TITLE
"Fehler bei Daten abrufen"  - Problems to fetch Data

### ISSUE THREAD

**Issue by thiemt** [deg0nz/MMM-PublicTransportBerlin#32]
I get the following error Message display at the Mirror:

"Fehler bei Daten abrufen: {"isHafasError":true,"message":"Internal Server Error","statusCode":500}
Das könnte heißen, dass VBB-Daten gerade nicht verfügbar sind."

I used the sample config, and also reinstalled everything. But even after waiting a few days I get this message. So I think it is not a due to  unavailability of the VBB Backend as the error message is intimating. Can you help me here?

**Comment by deg0nz** [comment]
Could you please try to update your module? The API endpoint changed and old versions do not support it anymore. Newer versions support the new endpoint.

**Comment by berlincount** [comment]
didn't help for me :( still not getting any data. latest git, latest npm modules, etc.

some commit missing?

**Comment by deg0nz** [comment]
hmm.. this is weird.
The problem is, I currently haven't set up my mirror, because i moved into a new flat... So I can't test stuff.
I'll check things on my mirror asap and will give you some feedback...

**Comment by deg0nz** [comment]
But actually, the current version with a fresh `npm install` should work (theoretically). I'm using it on my mirror for some time now... I Implemented to use `vbb-client` > 3.0.0 some months ago as mentioned in #27 .

Some people in the MagicMirror forum also had strange behaviour updating this module in the past... Have you tried a new install with deleting the whole `MMM-PublicTransportBerlin` folder?

**Comment by floge77** [comment]
Same here for me. Any recommendations how to Debug? Would like to help with debugging...

**Comment by deg0nz** [comment]
I'll try to turn on my mirror tonight and look into this when I get home...

@floge77 
you could write a small test-program which checks if the [Fetcher-Prototype](https://github.com/deg0nz/MMM-PublicTransportBerlin/blob/master/VbbFetcher.js) is working correctly. You just have to call the `fetchDepartures` function and see, if it is working correctly. This one:
https://github.com/deg0nz/MMM-PublicTransportBerlin/blob/390b60d65a5749bd4c0844789ece3a097d6e4880/VbbFetcher.js#L18

Theres also a `printDeparture` function that prints the departures more or less nicely to the console for debugging... Just iterate the response from `fetchDepartures` and put each line in this function.
https://github.com/deg0nz/MMM-PublicTransportBerlin/blob/390b60d65a5749bd4c0844789ece3a097d6e4880/VbbFetcher.js#L110

If anyone is familiar with wireshark/tcpdump, you could check, where the requests go to. The Wrong API is `vbb.transport.rest` (which actually shows what you are seeing on the mirror) the right one would be `2.vbb.transport.rest`.

I believe the problem is, that the module somehow is still using an old version of `vbb-client` in all or most of your cases...

**Comment by berlincount** [comment]
Hm. After checking myself, the right `vbb-client` is being used, and the right endpoint is being called. It's just that that endpoint is fscked ...

```
$ curl -v 'https://2.vbb.transport.rest/stations/9007102/departures?when=1529416634&duration=30'
*   Trying 54.37.75.136...
* Connected to 2.vbb.transport.rest (54.37.75.136) port 443 (#0)
* found 46 certificates in /etc/ssl/certs/ca-certificates.crt
* found 188 certificates in /etc/ssl/certs
* ALPN, offering http/1.1
* SSL connection using TLS1.2 / ECDHE_RSA_AES_256_GCM_SHA384
*        server certificate verification OK
*        server certificate status verification SKIPPED
*        common name: 2.vbb.transport.rest (matched)
*        server certificate expiration date OK
*        server certificate activation date OK
*        certificate public key: RSA
*        certificate version: #3
*        subject: CN=2.vbb.transport.rest
*        start date: Thu, 24 May 2018 21:10:15 GMT
*        expire date: Wed, 22 Aug 2018 21:10:15 GMT
*        issuer: C=US,O=Let's Encrypt,CN=Let's Encrypt Authority X3
*        compression: NULL
* ALPN, server accepted to use http/1.1
> GET /stations/9007102/departures?when=1529416634&duration=30 HTTP/1.1
> Host: 2.vbb.transport.rest
> User-Agent: curl/7.47.0
> Accept: */*
>
< HTTP/1.1 504 Gateway Time-out
< Server: nginx/1.10.3 (Ubuntu)
< Date: Tue, 19 Jun 2018 13:53:00 GMT
< Content-Type: text/html
< Content-Length: 192
< Connection: keep-alive
<
<html>
<head><title>504 Gateway Time-out</title></head>
<body bgcolor="white">
<center><h1>504 Gateway Time-out</h1></center>
<hr><center>nginx/1.10.3 (Ubuntu)</center>
</body>
</html>
* Connection #0 to host 2.vbb.transport.rest left intact
```

No idea how to solve that.

**Comment by berlincount** [comment]
I've opened https://github.com/derhuerst/vbb-rest/issues/29

**Comment by deg0nz** [comment]
I had a look on [https://status.transport.rest/](https://status.transport.rest/) and the endpoint seems to be down the whole day today...

**Comment by berlincount** [comment]
Must've been down for _days_, then .. or the monitor is giving false negatives.

**Comment by Juliadtc** [comment]
hey guys, has someone fixed the problem? I am stuck with: error while fetching departures: Error: Gateway Time-out for a couple of days now and cant find any solution

**Comment by rejas** [comment]
like @deg0nz wrote: the api endpoint at transport.rest seems to be down. no chance for this module to get data from there, thats why we get errors.

**Comment by Juliadtc** [comment]
yep, any experience how long such a downer will last? Thanks for your answer

**Comment by rejas** [comment]
no experience from my side, your best bet is to follow this issue: https://github.com/derhuerst/vbb-rest/issues/29

**Comment by philipppfrank** [comment]
Same problem here.

**Comment by deg0nz** [comment]
The outage of the API has never been so long before. So please check https://status.transport.rest once in a while. If the problem still exists when the API is available again, I'll look into the issue later.

I'll be off the grid for about two weeks from Tuesday on, so I kindly ask you all to be patient till mid July :)
If there are still errors, pull requests are more than welcome ;)

**Comment by rejas** [comment]
I will keep an eye on it (this issue, the other issue, the endpoint) enjoy your off-grid-time!

**Comment by rejas** [comment]
Endpoint should be reachable again. 
To everyone using this module: consider supporting the provider of the API @derhuerst (for details look into the ticket https://github.com/derhuerst/vbb-rest/issues/29#issuecomment-401064247)

**Comment by berlincount** [comment]
Hm. Maybe this should be a sidecar to deploy locally, i.e. everyone should be able to run the VBB REST translation service locally instead of depending on an external third party ...

**Comment by derhuerst** [comment]
Author of `vbb-rest`, the API behind `2.vbb.transport.rest` here.

> ```
> Error while fetching departures (for Station ID 9009203): Error: Bad Gateway
> Error while fetching departures (for Station ID 9007102): Error: Bad Gateway
> ```

A [`504` error](https://en.wikipedia.org/wiki/List_of_HTTP_status_codes#5xx_Server_errors) indicates that the VBB API, which the `2.vbb.transport.rest` wraps, returned an error. This is something I can't fix.

> If anyone is familiar with wireshark/tcpdump, you could check, where the requests go to. The Wrong API is `vbb.transport.rest` (which actually shows what you are seeing on the mirror) the right one would be `2.vbb.transport.rest`.
>
> I believe the problem is, that the module somehow is still using an old version of `vbb-client` in all or most of your cases...

Users can actually check this by inspecting `node_modules/vbb-client`. The `version` in `node_modules/vbb-client/package.json` should start with `3` and the [`endpoint` variable in `node_modules/vbb-client/lib/request.js`](https://github.com/derhuerst/vbb-client/blob/c162e4aece867c9c3b3f78034d0a3ec38a81a562/lib/request.js#L7) should contain `2.vbb.transport.rest`.

Nevertheless, they should all update once in a while to make it easier for me to turn off old endpoints and features, as well as introduce new ones.

> I had a look on https://status.transport.rest/ and the endpoint seems to be down the whole day today...

That's right. [As mentioned in derhuerst/vbb-rest#29](https://github.com/derhuerst/vbb-rest/issues/29#issuecomment-398671517), apparently the upstream VBB API has blocked the IP address of my server. I didn't have time to deal with it properly. **Luckily, @juliuste is hosting the endpoint now and its up again!** ❤️

If have ideas or resources to make the server more reliable/resilient, participate in derhuerst/vbb-rest#29.

> Hm. Maybe this should be a sidecar to deploy locally, i.e. everyone should be able to run the VBB REST translation service locally instead of depending on an external third party ...

**This is indeed a good idea if you depend on it for critical things!** (The upstream VBB API is not too reliable either, though.) Or use the underlying lib [`vbb-hafas`](https://github.com/derhuerst/vbb-hafas) directly if you use JS.

For all other magic mirrors and quick hack projects, this is not feasible. The goal is that, eventually, public transport providers will host a proper API themselves. Until then, we need proxies like `2.vbb.transport.rest`.

**Comment by ferby09** [comment]
Unfortunately it happened again...

**Comment by derhuerst** [comment]
Yup, working on a mitigation. Follow derhuerst/vbb-rest#29 for more details.

It might be the case that there's no long-term solution and we have to shut down the API though.

**Comment by ferby09** [comment]
I just cloned the repo from scratch including commit 11a2f93, but I'm getting "Fehler bei Daten abrufen" again. I even changed the fetch url in node_modules/vbb-client/lib/request.js to a mirror hosted on now.sh, because it did the trick the last couple of days but to no avail.

**Comment by deg0nz** [comment]
Yeah, I have that error too, but it does not come from upstream I guess. I'm investigating...

@ferby09 Is your error object on the mirror empty? (showing "{}"?)

**Comment by deg0nz** [comment]
@ferby09 I did a workaround for the error I had with 6f935b2. Could you please pull the current version and check, if the problem persists?

**Comment by derhuerst** [comment]
The repo I asked people to deploy/mirror is not the regular HTTP API (`vbb-rest`), but a WebSockets API (`hafas-client-rpc`). Therefore you can't change the the magic mirror to point to the WebSockets server.

**If you want to use your own HTTP API (`vbb-rest`) server, you're welcome to do so. Less requests on `2.vbb.transport.rest` will reduce the change of for it to get blocked again.** But if you cannot host a nserver or don't want to invest the time, you can use `2.vbb.transport.rest`.

**Comment by deg0nz** [comment]
Yeah, I understood that :)

Maybe I could change the setup for the mirror module to run the `vbb-rest` server locally on the mirror machines. Since some people are using the module now, that could make sense... I'll think about that.

But as I see at first sight, I could also change the module to use `hafas-client-rpc` instead of `vbb-client` though...

**Comment by derhuerst** [comment]
If you can run all Node modules on a magic mirror, use `vbb-hafas`, `hafas-client`! It directly queries HAFAS. We should make the user agent mandatory, for the VBB to be able to tell who causes traffic.

**Comment by deg0nz** [comment]
Yeah, you can run any Node module on a mirror. I'll put this on my roadmap to use `vbb-hafas`/`hafas-client` locally. I hope, the new system with load balancing will withstand until then...

> We should make the user agent mandatory, for the VBB to be able to tell who causes traffic.

Could you clarify what you mean by that? Do you mean we should set the user agent of a locally running `vbb-hafas` instance to something that identifies the MagicMirror?

**Comment by derhuerst** [comment]
>> We should make the user agent mandatory, for the VBB to be able to tell who causes traffic.
> 
> Could you clarify what you mean by that? Do you mean we should set the user agent of a locally running vbb-hafas instance to something that identifies the MagicMirror?

Yes. I should make a user agent string mandatory in `hafas-client`. You should, if possible, ask for a descriptive user agent like "alex's magic mirror".

This will

- reduce the number of requests with `derhuerst` in the user agent (VBB complained about that, but if they see it's not all me, but other people, they might understand),
- help them lock out *individual* clients that act abusive by sending too many requests.

**Comment by ferby09** [comment]
I did a clean clone of the repo again and even restarted my pi but the error persists.
{"isHafasError":true,"message":"Internal Server Error","statusCode":500}

**Comment by derhuerst** [comment]
@ferby09 I need more info to be able to reproduce.

**Comment by rejas** [comment]
I get this error too, but only when I use the field "directionStationId"
Otherwise the data flows correctly unto my mirror.
So maybe that direction-Feature is buggy (as it was mentioned in the README).

**Comment by ferby09** [comment]
I'm using directionStationId as well and as @rejas suspected the error disappeared when I commented this out. Unfortunately it is the most important feature for me as I'd have way too much irrelevant output without it.

**Comment by derhuerst** [comment]
Yup, it's buggy. Follow derhuerst/vbb-rest#31. I might be able to fix it tomorrow.

**Comment by derhuerst** [comment]
As you might have noticed, the endpoint is down again because VBB has blocked the new nodes as well. **I won't run the public endpoint at `2.vbb.transport.rest` for the time being.** Maybe VBB and I will find a compromise when we meet next month. Refer to derhuerst/vbb-rest#29 for more details.

If you want to use consume VBB data using JavaScript, use [`vbb-hafas`](https://github.com/derhuerst/vbb-hafas) directly, otherwise host your own `vbb-rest` instance for now.

To quote myself from above:

>> Hm. Maybe this should be a sidecar to deploy locally, i.e. everyone should be able to run the VBB REST translation service locally instead of depending on an external third party ...
>
> **This is indeed a good idea if you depend on it for critical things!** (The upstream VBB API is not too reliable either, though.) Or use the underlying lib [`vbb-hafas`](https://github.com/derhuerst/vbb-hafas) directly if you use JS.
>
> For all other magic mirrors and quick hack projects, this is not feasible. The goal is that, eventually, public transport providers will host a proper API themselves. Until then, we need proxies like `2.vbb.transport.rest`.

**Comment by berlincount** [comment]
I wasn't able to convert `MMM-PublicTransportBerlin` to `vbb-hafas` correctly - can someone lend a hand? .. I think polling individually might solve a lot of problems, but this might want a configurable option?

**Comment by deg0nz** [comment]
@berlincount What do you exactly mean by "but this might want a configurable option"?

Configure the module to use either `vbb-client` or `vbb-hafas`?

I successfully got `hafas-client` working on my mirror with the module. Gonna push the changes to master in the next hour. Let's see how long this works...

**Comment by berlincount** [comment]
@deg0nz yeah. but I think using HAFAS directly is probably the best option either way :-/

**Comment by derhuerst** [comment]
@deg0nz Are you using `hafas-client@next`?

**Comment by deg0nz** [comment]
@derhuerst No, I just installed the current version of `vbb-hafas` via npm. What are the advantages to use `hafas-client@next`?
It would be no problem to convert it to `hafas-client` with the VBB profile for me though...

**Comment by deg0nz** [comment]
@berlincount 

>yeah. but I think using HAFAS directly is probably the best option either way :-/

Yeah, I think so too... We'll just leave it as it is for now... 
I'll just have to implement the mandatory user-agent stuff for future versions of `hafas-client` as it was added in https://github.com/public-transport/hafas-client/commit/5d9d738152e60855388e22e9d1b88b5ca21d4c37 and will be feature of the next version.

**Comment by berlincount** [comment]
I'm happily seeing trains again on my mirror (with Ringbahn showing train numbers instead of icons, though), so .. it's in a useful state again for me.

My two cents: I'd love the User-Agent to be something helpful, e.g. `MagicMirror/2.4.1 (x86_64; de-de) Electron/2.0.6 node/6.12.2 MMM-PublicTransportBerlin/1.2.4 vbb-hafas/4.2.1 hafas-client/2.3.2` over something `berlincount's MagicMirror 1.2.3` for my current setup ... but people might have different opinions or that, or want configurability. Maybe this even should be solved in a MagicMirror-global way, so your sub-modules can inform the overall `requests` systems (which sets the general info) what to report upstream additionally?

**Comment by derhuerst** [comment]
> My two cents: I'd love the User-Agent to be something helpful, e.g. `MagicMirror/2.4.1 (x86_64; de-de) Electron/2.0.6 node/6.12.2 MMM-PublicTransportBerlin/1.2.4 vbb-hafas/4.2.1 hafas-client/2.3.2` over something `berlincount's MagicMirror 1.2.3` for my current setup ... but people might have different opinions or that, or want configurability.

What is your motivation to send such a `user-agent` to the VBB/BVG API?

I made the `user-agent` parameter mandatory (with the next major version of `hafas-client`) in order to show VBB that people use the API for their projects, while making clear that it is *not me* who send the requests.

**Comment by berlincount** [comment]
Uhm, because that's the definition of `User-Agent` per https://tools.ietf.org/html/rfc2616?

```
3.8 Product Tokens

   Product tokens are used to allow communicating applications to
   identify themselves by software name and version. Most fields using
   product tokens also allow sub-products which form a significant part
   of the application to be listed, separated by white space. By
   convention, the products are listed in order of their significance
   for identifying the application.

       product         = token ["/" product-version]
       product-version = token

   Examples:

       User-Agent: CERN-LineMode/2.15 libwww/2.17b3
       Server: Apache/0.8.4

   Product tokens SHOULD be short and to the point. They MUST NOT be
   used for advertising or other non-essential information. Although any
   token character MAY appear in a product-version, this token SHOULD
   only be used for a version identifier (i.e., successive versions of
   the same product SHOULD only differ in the product-version portion of
   the product value).
```

Differing software versions would also be an indicator of different users, AND a clear indication to VBB what kind of software is using their service for what perpose would allow for better planning on their side, especially as OpenSource Software's characteristics can be better understood when you know what you're dealing with - and in which version (i.e. pre- or post-bug, before a certain optimization, etc ..).

I've seen setups where different client versions (determined by `User-Agent`) where load-balanced to different servers clusters with specific optimizations.

I think the use-case you're talking about would much more be suited to an `X-Forwarded-For`, as commonly used and later discussed in https://tools.ietf.org/html/rfc2616

```
A common way to disclose this information is by using the non-
   standard header fields such as X-Forwarded-For, X-Forwarded-By, and
   X-Forwarded-Proto.  There are many benefits to using a standardized
   approach to commonly desired protocol function: not least is
   interoperability between implementations.  
```

Standardized configuration options exist for common server solutions like Nginx https://ma.ttias.be/nginx-access-log-log-the-real-users-ip-instead-of-the-proxy/ or Apache http://www.techstacks.com/howto/log-client-ip-and-xforwardedfor-ip-in-apache.html to use this header.

Proper use of `X-Forwarded-For` also allows tracing abuse back to the actual source, as well as allowing more meaningful statistics about e.g. Geolocation of the users - the same goes for `User-Agent`, which can be used to for statistics about software versions used.

Forcing users to set any values uniquely is also a usuability nightmare, as it's tempting to either just put things like "foobar", "user-agent" or "Martin's Mirror", which .. might not be unique. The use of a UUID might also be counterindicated, and cause more privacy issues than forwarding the actual user IP (which I'd be exposing anyway if I weren't using your convenience-proxy .. the average user won't care about the actual implementation or hidden interface definitions) ...

Am I making sense? :)

**Comment by derhuerst** [comment]
>>> I'd love the User-Agent to be something helpful, e.g. `MagicMirror/2.4.1 (x86_64; de-de) Electron/2.0.6 node/6.12.2 MMM-PublicTransportBerlin/1.2.4 vbb-hafas/4.2.1 hafas-client/2.3.2` over something `berlincount's MagicMirror 1.2.3` [...].
>>
>> What is your motivation to send such a user-agent to the VBB/BVG API?
>
> Uhm, because that's the definition of User-Agent per https://tools.ietf.org/html/rfc2616?

Okay, then I have misunderstood you. Obviously @deg0nz can adapt `MMM-PublicTransporBerlin` to pass a very specific `User-Agent` into `hafas-client`. Fine with me!

> Differing software versions would also be an indicator of different users, AND a clear indication to VBB what kind of software is using their service for what perpose would allow for better planning on their side [...].

While that is true theoretically (and most likely HaCon analyses the distribution of their app versions), in practice they block clients based on the `User-Agent`. Providing a more detailed (hence more unique) one will only increase your chance of getting blocked.

In fact, the original motivation so send `https://github.com/public-transport/hafas-client` as `User-Agent` was to communicate that

- there is a community out there, making use of public transport info, that would also benefit from a truly open API,
- the community runs open source software to power their hacks/gadgets.

> I think the use-case you're talking about would much more be suited to an `X-Forwarded-For`, as commonly used and later discussed in https://tools.ietf.org/html/rfc2616

HTTP proxies usually send this header, so I don't want to interfere with that.

Or do you mean that I should send this header from `*.transport.rest`? I definitely agree with that, PR welcome!

> Forcing users to set any values uniquely is also a usuability nightmare, as it's tempting to either just put things like "foobar", "user-agent" or "Martin's Mirror", which .. might not be unique. The use of a UUID might also be counterindicated, and cause more privacy issues than forwarding the actual user IP (which I'd be exposing anyway if I weren't using your convenience-proxy .. the average user won't care about the actual implementation or hidden interface definitions) ...

I do care about usability and privacy, don't get me wrong, but this is about being able to query the API in the first place. It seems like all requests containing `hafas-client` are being blocked by now. [They're already blocking based on IP addresses](
https://github.com/deg0nz/MMM-PublicTransportBerlin/issues/32#issuecomment-406795157).

> Am I making sense? :)

Yup.

**Comment by derhuerst** [comment]
For the latest updates on their blocking efforts, see https://github.com/derhuerst/vbb-rest/issues/29#issuecomment-411469242.

**Comment by deg0nz** [comment]
Hi all,

since the BVG endpoint seems to work for some days now (see https://status.transport.rest), I changed the module to use `hafas-client` with the BVG profile with commit 09bee5c. Let's see how long this will work...

@derhuerst How are your experiences with VBB-Stations in Brandenburg with the BVG profile? Are they still supported like in the VBB profile?

**Comment by derhuerst** [comment]
Living in Berlin, I have a strong Berlin bias, but from my experiences, the VBB station IDs should work.

You can test it by looping through [`vbb-stations`](https://npmjs.com/package/vbb-stations) and using `hafas-client.stop(id)` to check if HAFAS knows about them.

**Comment by deg0nz** [comment]
This is true for me as well (the Berlin bias). I'll try the Brandenburg stations when I find time to do so.

**Comment by deg0nz** [comment]
So the module is working fine with the BVG-Profile for almost a month now. I will close this issue.
If you encounter any problems like this feel free to open it again (or create a new issue).
